content/courses/ada-idioms/chapters/reducing_object_code_from_generic_package_instantiations.rst

```diff
@@ -0,0 +1,169 @@
+package :ada:`P` be a distinct package, as if each instance was instead written
+explicitly as a non-generic package (at the point of instantiation) with the
+instance name. As a result, each package instance declares an exception, and
+these exceptions must be treated as distinct from each other. A code-sharing
+implementation must maintain that distinction with one object code module.
+
+In the example above, there are no generic formal parameters, nor other
+declarations within the package declarative part besides the exception, because
+they are not necessary for that example.  However, generic formal parameters
+can be a problem for code-sharing too. For example, consider this generic
+package:
+
+.. code-block:: ada
+
+    generic
+       Formal_Object : in out Integer;
+    package P is
+       --  ...
+    end P;
+
+This generic package has a generic formal object parameter with mode
+:ada:`in out`. (We chose type :ada:`Integer` purely for convenience.)  That
+specific mode can cause a similar problem as seen in the exception example,
+because the mode allows the generic package to update the generic actual
+object passed to it. The shared object code must keep track of which object is
+updated during execution.
+
+Therefore, when writing the application source code that instantiates generic
+packages, developers should do so in a manner that minimizes the amount of
+object code that might result.
+
+
+Solution
+--------
+
+The application source code should be written in a manner that shares the
+instantiations themselves, when possible, thereby reducing the number of
+instantiations that exist.
+
+For example, let’s say that several units in the application code require the
+ability to do I/O on some floating-point type. For simplicity, let’s say that
+this is a type named :ada:`Real`, declared in a package named :ada:`Common`.
+Here is a declaration for an example package body that requires the I/O
+capability:
+
+.. code-block:: ada
+
+    with Ada.Text_IO, Common;
+    package body User1 is
+        package Real_IO is new Ada.Text_IO.Float_IO (Common.Real);
+        -- ...
+    end User1;
+
+That’s certainly legal, and works, but we’ve said that several units require
+I/O for type :ada:`Real`. Let’s say there are in fact twenty such units. They
+all do something similar:
+
+.. code-block:: ada
+
+    with Ada.Text_IO, Common;
+    package body User20 is
+        package Real_IO is new Ada.Text_IO.Float_IO (Common.Real);
+        --  ...
+    end User20;
+
+As a result, the application has twenty instantiations (at least) of
+:ada:`Ada.Text_IO.Float_IO`. There will be instances named
+:ada:`User1.Real_IO`, :ada:`User2.Real_IO`, and so on, up to
+:ada:`User20.Real_IO`. The fact that the local names are all
+:ada:`Real_IO` is irrelevant.
+
+If the compiler happens to use the macro-expansion implementation, that means
+the application executable will have twenty copies of the object code defined
+by the generic :ada:`Float_IO`.
```

Comment:
"by the generic :ada:`Float_IO`." => "by the generic :ada:`Float_IO`. GNAT performs some internal restructuring to avoid this problem for these specific language-defined generic units, but not for application-defined generics."